### PR TITLE
Implement adding comment and auto assigning to relevant reviewers

### PR DESCRIPTION
**Pull request by tapishr**
Please check the action items covered in the PR - 

- [x] Build is running
- [x] Manual QA
- [x] Please review PR #5 before reviewing this PR
 
In this PR, following flow is covered -> 
-> User pushes to PR
-> next server gets webhook callback message, passes to dpu in pubsub
-> dpu receives pubsub message in listener.rs and fires up a task to process it
-> dpu calculates hunks
----------<>Everything upto this point is already implemented, below steps are new in this pr<>-----------------
-> from hunks, dpu creates comment message and adds it
-> dpu adds relevant reviewers if they are present in sled db (are a member of the workspace) and not already added to the PR

### Summary by CodeRabbit

- New Feature: Added authentication and token management for Bitbucket integration.
- New Feature: Implemented functionality to add comments and reviewers to Bitbucket pull requests.
- New Feature: Added ability to retrieve and store workspace repositories from Bitbucket.
- New Feature: Implemented functionality to manage workspace users in the database.
- New Feature: Added support for webhooks, including retrieval and addition of webhooks in a repository.
- New Feature: Implemented code coverage analysis and review comments.
- New Feature: Added functionality to process and store review data.
- New Feature: Enhanced setup process with new logic for retrieving access tokens and workspace information from Bitbucket.
- New Feature: Implemented PubSub message processing.
- New Feature: Added new data structures for handling repositories, reviews, users, and webhooks.
- Improvement: Enhanced error handling and logging in various functions.
- Refactor: Codebase has been refactored into multiple modules for better organization and maintainability.

### CODE REVIEW

File: vibi-dpu/src/bitbucket/reviewer.rs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+use std::io;
+
+use crate::db::user::get_workspace_user_from_db;
+use crate::utils::review::Review;
+use crate::utils::user::BitbucketUser;
+
+use super::{config::{bitbucket_base_url, get_client, prepare_headers}};
+
+pub async fn add_reviewers(user_key: &str, review: &Review, access_token: &str) {
+    let url = prepare_add_reviewers_url(review.repo_owner(), review.repo_name(), review.id());
```
Comment (tapishr) [inline vibi-dpu/src/bitbucket/reviewer.rs:14]: This function is not correctly named. It only gets the pr info, doesn't add reviewers

Comment (avikalpg) [inline vibi-dpu/src/bitbucket/reviewer.rs:14]: incorrect function name.

File: vibi-dpu/src/bitbucket/reviewer.rs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    let get_response = get_pr_info(&url, access_token).await;
+    let put_request_body_opt = add_user_to_reviewers(get_response, user_key).await;
```
Comment (avikalpg) [inline vibi-dpu/src/bitbucket/reviewer.rs:16]: This is the actual `prepare_add_reviewers_url` function.

File: vibi-dpu/src/bitbucket/reviewer.rs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    // For each user in user_from_db.users()...
+    for user in user_from_db.users().iter() {
+        // If the reviewers vector doesn't contain the user...
+        if !reviewers.contains(user) {
+            // Add the user to reviewers
+            reviewers.push(user.clone());
+        }
+    }
+    println!("Updated reviewers = {:?}", reviewers);
+    let payload_opt = prepare_put_body(&reviewers, &get_response_json);
```
Comment (tapishr) [inline vibi-dpu/src/bitbucket/reviewer.rs:45]: prepare_put_body and put_call should be together

File: vibi-dpu/src/bitbucket/reviewer.rs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    println!("put reviewers payload = {:?}", &payload_opt);
+    return payload_opt;
+}
+
+fn prepare_put_body(updated_reviewers: &Vec<BitbucketUser>, get_response_json: &Value) -> Option<Value> {
```
Comment (tapishr) [inline vibi-dpu/src/bitbucket/reviewer.rs:50]: get_response_json to be replaced by get_existing_pr_info

File: vibi-dpu/src/core/coverage.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
+use std::collections::{HashMap, HashSet};
+
+use crate::{utils::hunk::{HunkMap, PrHunkItem}, db::user::get_workspace_user_from_db, bitbucket::{comment::add_comment, reviewer::add_reviewers}};
+use crate::utils::review::Review;
+use crate::bitbucket::auth::get_access_token_review;
+
+pub async fn process_coverage(hunkmap: &HunkMap, review: &Review) {
+    let access_token_opt = get_access_token_review(review).await;
```
Comment (tapishr) [inline vibi-dpu/src/core/coverage.rs:8]: Why is this called

File: vibi-dpu/src/core/coverage.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    if access_token_opt.is_none() {
+        eprintln!("Unable to acquire access_token in process_coverage");
+        return;
+    }
+    let access_token = access_token_opt.expect("Empty access_token_opt");
+    for prhunk in hunkmap.prhunkvec() {
+        // calculate number of hunks for each userid
+        let coverage_map = calculate_coverage(&hunkmap.repo_owner(), prhunk);
+        if !coverage_map.is_empty() {
```
Comment (tapishr) [inline vibi-dpu/src/core/coverage.rs:17]: Implement never nesting

File: vibi-dpu/src/core/coverage.rs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            coverage_floatmap.insert(author_id, coverage);
+        }
+        else {
+            coverage_floatmap.insert(author_id, num_lines);
+        }
+    }
+    if total <= 0.0 {
+        return coverage_map;
+    } 
+    for (key, value) in coverage_floatmap.iter_mut() {
```
Comment (tapishr) [inline vibi-dpu/src/core/coverage.rs:61]: key -> blame_author, value -> coverage

File: vibi-dpu/src/core/coverage.rs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        if coverage_floatmap.contains_key(&author_id) {
+            let coverage = coverage_floatmap.get(&author_id).expect("unable to find coverage for author")
+                + num_lines;
+            coverage_floatmap.insert(author_id, coverage);
+        }
+        else {
+            coverage_floatmap.insert(author_id, num_lines);
+        }
+    }
+    if total <= 0.0 {
```
Comment (tapishr) [inline vibi-dpu/src/core/coverage.rs:58]: declare coverage_map here

File: vibi-dpu/src/core/review.rs [hunk 9]
```diff
@@ -41,8 +42,9 @@ fn send_hunkmap(hunkmap_opt: &Option<HunkMap>, review: &Review) {
 	println!("HunkMap = {:?}", &hunkmap);
 	store_hunkmap_to_db(&hunkmap, review);
 	publish_hunkmap(&hunkmap);
-	// let hunkmap_async = hunkmap.clone();
-	// process_coverage(&hunkmap_async).await; TODO - include in future PR
+	let hunkmap_async = hunkmap.clone();
+	let review_async = review.clone();
+	process_coverage(&hunkmap_async, &review_async).await;
```
Comment (avikalpg) [inline vibi-dpu/src/core/review.rs:47]: We can add a status column to the hunks table in the database which we can update based on the success of this function for easier status report to the client.

**Review by avikalpg — APPROVED** [review]

File: vibi-dpu/src/core/coverage.rs (hunk 5 above)
Comment (tapishr) [reply vibi-dpu/src/core/coverage.rs:8]: Needs new PR

File: vibi-dpu/src/core/coverage.rs (hunk 6 above)
Comment (tapishr) [reply vibi-dpu/src/core/coverage.rs:17]: After PR for email/display name fix

**Pull request merged**